lib/cadet/accounts/luminus.ex

```diff
@@ -0,0 +1,234 @@
+      {:ok, "12742174091894830298409823098098"}
+  """
+
+  def fetch_luminus_token_or_return_default(nil) do
+    "token"
+  end
+
+  def fetch_luminus_token_or_return_default(code) do
+    case fetch_luminus_token(code) do
+      {:error, :bad_request} -> "token"
+      {:ok, token} -> token
+    end
+  end
+
+  def fetch_luminus_token(code) do
+    queries =
+      URI.encode_query(%{
+        client_id: @client_id,
+        client_secret: @client_secret,
+        grant_type: "authorization_code",
+        code: code,
+        redirect_uri: @redirect_url
+      })
+
+    headers = [{"Content-type", "application/x-www-form-urlencoded"}]
+
+    case HTTPoison.post(@api_token_url, queries, headers) do
+      {:ok, %{body: body, status_code: 200}} ->
+        {:ok, body |> Jason.decode!() |> Map.get("access_token")}
+
+      {:ok, %{status_code: 400}} ->
+        # LumiNUS responds with 400 if APIKey is invalid
+        {:error, :bad_request}
+    end
+  end
+
+  @doc """
+  Get the NUSNET ID and the name of the user corresponding to this token.
+
+  returns...
+
+    - {:ok, nusnet_id, name} - valid token, nusnet_id are name are strings
+    - {:error, :bad_request} - invalid token or luminus_client_secret is invalid
+
+  ## Parameters
+
+    - token: String, the LumiNUS authentication token
+
+  ## Examples
+
+      iex> Cadet.Accounts.LumiNUS.fetch_details("T0K3N...")
+      {:ok, "e012345", "LIOW JIA CHEN"}
+
+  """
+  def fetch_details(token) do
+    case api_call("user/Profile", token: token) do
+      {:ok, %{"userID" => nusnet_id, "userNameOriginal" => name}} -> {:ok, nusnet_id, name}
+      _ -> {:error, :bad_request}
+    end
+  end
+
+  @doc """
+  Get the role of the user corresponding to this token.
+
+  We check the end date of the module to ensure student is currently taking or teaching CS1101S
+  Roles:
+    student permission -> :student
+    manager / read manager permissions -> :staff
+    owner / co-owner -> :admin
+
+  returns...
+
+    - {:ok, :student} - valid token, has student permissions
+    - {:ok, :staff} - valid token, has manager or read manager permissions
+    - {:ok, :admin} - valid token, has owner or co-owner permissions
+    - {:error, :forbidden} - valid token, user does not currently read cs1101s
+    - {:error, :bad_request} - invalid token or luminus_client_secret is invalid
+
+  ## Parameters
+
+    - token: String, the LumiNUS authentication token
+
+  ## Parameters
+
+    - token: String, the LumiNUS authentication token
+
+  ## Examples
+
+      iex> Cadet.Accounts.Luminus.fetch_role("T0K3N...")
+      {:ok, :student}
+  """
+
+  def fetch_role(token) do
+    case api_call("module", token: token) do
+      {:ok, modules} ->
+        parse_modules(modules)
+
+      {:error, _} ->
+        {:error, :bad_request}
+    end
+  end
+
+  defp moduleActive?(endDate) do
+    Timex.before?(Timex.now(), Timex.parse!(endDate, "{ISO:Extended}"))
+  end
+
+  defp cs1101s?(name) do
```

Comment:
Why does this need to be a function? I think you should inline this. It's clearer that way.

Reply:
its about time we externalize this into the config file so we can use XS1101S for staging and dev purposes! haha another PR!